django/urls/converters.py

```diff
@@ -0,0 +1,73 @@
+import uuid
+
+from django.utils import lru_cache
+
+
+class IntConverter:
+    regex = '[0-9]+'
+
+    def to_python(self, value):
+        return int(value)
+
+    def to_url(self, value):
+        return str(value)
+
+
+class StringConverter:
+    regex = '[^/]+'
+
+    def to_python(self, value):
+        return value
+
+    def to_url(self, value):
+        return value
+
+
+class UUIDConverter:
+    regex = '[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}'
+
+    def to_python(self, value):
+        return uuid.UUID(value)
+
+    def to_url(self, value):
+        return str(value)
+
+
+class SlugConverter(StringConverter):
+    # Django has both `slug_re` and `slug_unicode_re`. In the case of URLs, the
+    # non-unicode variant seems to make more sense.
+    regex = '[-a-zA-Z0-9_]+'
+
+
+class PathConverter(StringConverter):
+    regex = '.+'
+
+
+DEFAULT_CONVERTERS = {
+    'int': IntConverter(),
+    'path': PathConverter(),
+    'slug': SlugConverter(),
+    'str': StringConverter(),
+    'uuid': UUIDConverter(),
+}
+
+
+REGISTERED_CONVERTERS = {}
+
+
+def register_converter(converter, typename):
+    REGISTERED_CONVERTERS[typename] = converter()
+    get_converters.cache_clear()
```

Comment:
According to the DEP, this should be usable as a class decorator, but I don't see this here...
Something like
```python
def register_converter(converter, typename=None):
    if typename is None and isinstance(converter, str):
        # We're used as a decorator
        return functools.partial(register_converter, typename=converter)
    else:
       ... # current body
```
Alternatively, make `typename` an attribute of the converter class -- then `register_converter` takes a single argument and can trivially be used as a decorator. Making the name accessible in the converter is probably better anyway, for error-reporting in any non-trivial `to_python()` or `to_url()` method.

Reply:
Based on your suggestion, I looked into making it available as a decorator (and pushed a half-working commit which I later removed again).

1. First of all, current body returns nothing, so that would need to be added.
2. When writing the tests for this, I was updating the Base64Converter to use the decorator version. Flake8 immediately started complaining about an unused import of `converters`.

Based on the writing of the tests, I'm feeling less certain about using it as a decorator, and would maybe suggest updating the DEP to remove that option.

Your alternative of making the typename available as an attribute is also an option: it would help in the error reporting. But that could also be solved by `register_converter` setting the value as an attribute on the converter *instance* instead of having to define it on the converter *class*. It still does not solve the "unused import", though.